### ISSUE TITLE
No via without a track

### ISSUE THREAD

**Issue by marekr** [marekr/kicad-issue-test#180]
PCBnew does not allow to place arbitrary vias, that is, I cannot add a via unless I'm drawing a track. This makes it impossible to stitch two ground planes on separate layers together. There are workarounds but they are too cumbersome to use in board designs where a large number of stitches is desired.

Judging from PCBnew's manual this is intended behaviour. It should, however, be considered as a misfeature as this is very basic functionality found in practically every other PCB design program.

**Comment by marekr** [comment]
(by l-marcantonio)
On Mon, May 14, 2012 at 10:02:41AM -0000, Tommi Karkkainen wrote:
> Judging from PCBnew's manual this is intended behaviour. It should,
> however, be considered as a misfeature as this is very basic
> functionality found in practically every other PCB design program.

Yes, it is intended since it is 'not obvious' which netlist to assign to
these via; I'd put it in the wishlist (there is some work in progress
about netlist to track assigment but I have no idea if it's related to
your question).

-- 
Lorenzo Marcantonio
Logos Srl

**Comment by marekr** [comment]
(by kvasina)
placing arbitrary vias works for me:
- select layer
- select wire tool
- click on zone or track
- type V to place via.
- check board

for example when dealing with high ampacity tracks, i ended placing multiple vias to create low resistance path between layers.

but there is an issue. any further work that requires hiding zones or reflowing etc isolates those arbitrary vias. more over, this also creates isolation gap around via further diminishing ampacity of the track and DRC does nothing to warn you about this.
when via is placed it should be assigned net name so reflowing etc do not affect it.

**Comment by marekr** [comment]
(by szymon123)
it would be nice to put a single pad in arbitrary position. I usually use this feature to add holes on the boards.

**Comment by marekr** [comment]
(by oxplot)
A bullet proof and headache free way right now is to:

- create a module with a single through-hole pad (you do this once and use it forever)
- in pcbnew, add this module (no need to add it to schematics)
- edit the pad of this module and set its netname to your the name of your ground plane (e.g. GND)
- make as many copies as you need

Advantages:

- you don't have to do any maintenance—no touching your schematics, no reconnecting to ground plane after re-filling zones, etc
- you can change the size of or any other properties (thermal relief, drill size, etc) of all the pads at once if needed (just right click any of them and select "Global Pads Edition" under Pads menu)

Disadvantages:

- no shortcut to add a netname configured pad—you need to copy one of the existing ones

**Comment by marekr** [comment]
(by mikk-leini)
One little problem with "headache free way" is that it leads to another headache - footprints copy-pasting is not so easy (bug report #1233586).

Could you please prioritize adding an abstract pads/vias feature to PCB editor? This seems to be the most obvious missing feature.

**Comment by marekr** [comment]
(by stefankraus)
This is really a nasty bug. The workaround that szymon described helps (I simply used a "wirepad" in the library), but has its quirks. First of all, the DRC should warn, if I place a part under another part (which I have to do if I want to place a "via" in the thermal pad of an IC) - for some reason, the DRC does not complain. It also does not allow me to use my preconfigured via sizes but I would have to make a part for every size.

In Altium, you can just place vias and if you place it on a polygon on the selected layer, it automaticly uses the net of the polygon. You can also change the net manually (in Kicad, you would just press 'e' on a via and get a menu, where you could select via size, net, and so on).

**Comment by marekr** [comment]
(by onvav)
Is there any progress in resolving of this issue? I think that it is important limitation of KiCad. Arbitrary vias are common practice. Please consider a solution.

**Comment by marekr** [comment]
(by zeilhofer)
I'm also missing this feature!

*) Problem now when using the CERN interactive router:
When I want to place heat-sink-vias on an exposed pad, the router always snaps to the center of the exposed pad. 
So the normal workaround by drawing traces alternating on top and bottom doesn't work here. 
*) Workaround:
Draw a trace from the desired net into an empty region. Ther you can draw the vias with the alternating top/bot-method. Then delete the drawn traces, and keep the vias. The vias can now be moved to the desired place. 
*) Remark: I did try the method with crating a "via-module", placing many "vias", and manually connect them to the desired net. This way is very time consuming, due to the manual connection to the net. 


Perhaps the procedure of a very simple implementation of the missing feature could be:
1) select place-vias-tool
2) a window pops up, where one can select the net out of a list
3) place the vias. 

I would donate €100, if someone will implement this feature!

**Comment by marekr** [comment]
(by orsonmmz)
Hi Karl,

Regarding the snapping problem - you may temporarily disable the snapping by holding Shift while routing.

Regards,
Orson

**Comment by marekr** [comment]
(by haunma)
Sorry to pile on, but I just wanted to add my voice to the chorus.  KiCad is really improving fast and I so much want to use it for my next project, but this is the one thing holding me back.  Frankly, I don't understand why the CERN folks are working on things like push-and-shove when this feature request is still open.  You can do a serious layout with primitive manual routing but you can't do a serious layout without lots of arbitrary vias, at least not if signal integrity is an issue.  Are there working EEs consulting on the prioritization?

Regarding the via-as-footprint solution proposed above, how do you keep the "via" covered with soldermask?  Can you set up the module/footprint so no copper is showing through?  Otherwise it doesn't seem like a practical workaround...

Mark

**Comment by marekr** [comment]
(by orsonmmz)
Hi Mark,

We hoped the Push and Shove router will significantly speed up the routing process. Via stitching is on our to-do list, and that means it will appear in KiCad one day. As far as I understand, you would like to place vias between polygons to e.g. improve grounding. It can be done, you just need to have a track connecting the vias. I agree it is not the most convenient way to solve the problem, but currently it is possible.
We have a long list of features we would like to implement in KiCad. Some of the tasks depend on others, and that is one of the major factors in prioritization. It is hard to make everyone happy with limited man power (please note that KICad is not the primary mission of CERN; there is a very limited budget, and we raise funds ourselves to keep going). If you read other comments you may notice many "why is XXX still not implemented? that's the most important feature in EDA!" and we agree with most of the people. Not to mention that everyone has an individual idea for a tool and they are not always converging.
How about joining our efforts to make KiCad better? Recently there are more and more contributors coming and I am really glad to see the project growing.

Regards,
Orson

**Comment by marekr** [comment]
(by haunma)
Thanks Orson,

Sadly I'm not in a position to donate code but I would donate some cash.  Is there some place to do that online?

I know everyone thinks "their bug" is the most important, but in this case I'm right... really!  :)   I am serious about the suggestion to put these proposed features in front of some working EEs, people who do layout for a living, and ask them to help you rank the priorities.  I realize no one who does this professionally is going to switch from Altium [yet], but I think you might be surprised what floats to the top in such an exercise.

The push-and-shove stuff is awesome, and is going to save a LOT of time in layout, but in my experience (using a build from last year), the via issue is more than just a time waster.  If I knew I could get the result I want, by spending 10x or 100x of the time it would take in another layout package, this would not be such a problem.  But when I tried to string the vias along a track as suggested, pcbnew would delete old vias somewhat unpredictably whenever I tried to move things around later.  It also would  not allow adjacent vias to touch, which is another show-stopper.  (E.g., often when placing your array of grounding vias under a QFN package, they will need to touch in order to satisfy the PCB house design rules and still have enough vias.)  For my next project I will need to use some vias in an LVDS diff pair, and I am guessing the code will insist on having the vias centered on the traces, which will not allow me to get the impedance matched correctly.  These are just a few of the problems, off the top of my head.  I have pretty much accepted the fact that the boards I create with KiCAD will be electrically inferior.

Basically, until vias are treated as first-class objects which can exist independently of any net, this is going to be a problem.  IMO it is one of the few things gEDA got right, i.e. they take it as a given that the vias, and all copper generally, will stay wherever the user placed them; then, they run a connection engine to figure out which nets are actually connected to one another, and whether or not this matches the netlist.  There may be a better way to do it, but it needs to allow equivalent flexibility w.r.t. via and copper polygon placement.

I am still curious if the footprint-as-via technique, mentioned above by Mansour, will let you cover the whole thing with soldermask.  That might be the best workaround in the short term.

**Comment by marekr** [comment]
(by nickoe)
@Mark: I will only comment on a fraction of your last comment. Regarding: "but when I tried to string the vias along a track as suggested, pcbnew would delete old vias somewhat unpredictably whenever I tried to move things around later. It also would not allow adjacent vias to touch, which is another show-stopper."

As of now, I can do thoe two things in GAL when I disable the "Remove redundant tracks" for the pns router. I can also overlap vias. I have not tested this extensively with ground planes and such, or how hard this is to make work on a real design.

**Comment by marekr** [comment]
(by mikk-leini)
Seems like nobody is going to implement stitching vias (no such clue in roadmap) so i would like to try implementing this feature. I'm on vacation currently so i have little bit of time.
How should this be organized? I just attach a patch? I guess in the end somebody needs to review and merge it to main tree anyway, so maybe i can get some support right away?
I managed to build Kicad on Windows and add a new menu item but i don't yet know how complex it goes so support or idea sharing with developers would be really nice to have.

**Comment by marekr** [comment]
(by nickoe)
@Mikk,  I think it is best to announce your plan on the list and align with the rest of the developers on how it should be made.
